state/runtime/evm/instructions.go

```diff
@@ -805,45 +805,51 @@ func opCallDataCopy(c *state) {
 func opReturnDataCopy(c *state) {
 	if !c.config.Byzantium {
 		c.exit(errOpCodeNotFound)
-
 		return
 	}
 
 	memOffset := c.pop()
 	dataOffset := c.pop()
 	length := c.pop()
 
+	// Check if the dataOffset is uint64, this will also check for overflow
 	if !dataOffset.IsUint64() {
 		c.exit(errReturnDataOutOfBounds)
-
 		return
 	}
 
-	// if length is 0, return immediately since no need for the data copying nor memory allocation
-	if length.Sign() == 0 || !c.allocateMemory(memOffset, length) {
+	// Check if sum of dataOffset and length overflows uint64
+	var end = new(big.Int)
+	end.Add(dataOffset, length)
```

Comment:
Nitpick (we can inline it, because expression isn't relatively simple) WDYT? 🙂 
```suggestion
	end := new(big.Int).Add(dataOffset, length)
```